Add ansible-ssh role to generate ssh config

fullyint opened the pull request:
* Run `ansible-playbook dev.yml` without vagrant commands and without manually adding dev vm to ssh config. Grabs latest `vagrant ssh-config`, sorting out port forwarding between multiple VMs.
* Pairs nicely with #313 to run cross-environment playbooks (syncing files/database), including dev.
* Now you can use ssh config "includes". Files listed in `ssh_config_files` are `cat`-ed into `roles/ansible-ssh/files/config`.
* Initial discussion in #288
* #295 "Dynamic development IP" would be a really nice accompaniment

Could someone test on Windows?	
* May have to adjust or comment out the `ssh_config_files` in `group_vars/all` (?)
* The `vagrant ssh-config` command that runs would typically not be available when Windows users run `dev.yml` inside the vm. This gives return code 127, which I catch with `vagrant_ssh_config.rc == 127`, then set the `host` and `connection`, like in [`windows.sh`](https://github.com/roots/trellis/blob/8a84639257bdeaceb3b440823e2ac78c5c17906a/windows.sh#L39).
* Are the parameters `--sudo --user=vagrant` necessary in `windows.sh`? They already exist in `dev.yml`.

**Caution.** This refers to the vagrant vm as `vagrant-vm` in the hosts file. For consistency, this PR renames vagrant's reference to the vm from `default` to `vagrant-vm`. This latter means the Vagrantfile will stop working for existing VMs named `default`. Users would have to comment out the line
```
config.vm.define 'vagrant-vm'
```
and destroy the vm, then uncomment the line and `vagrant up`.

We don't have to rename the VM but it might be nice to have the names be consistent. Changing the name is a breaking change of sorts, but it seems better to do it now, while pre v1.0.

austinpray commented:
This fails for me:

```

PLAY [Prepare Ansible SSH Config] *********************************************

TASK: [ansible-ssh | Remove prior ssh config info] ****************************
ok: [vagrant-vm -> 127.0.0.1]

TASK: [ansible-ssh | Gather ssh config info for Vagrant VM] *******************
ok: [vagrant-vm -> 127.0.0.1]

TASK: [ansible-ssh | Gather user-specified ssh config files] ******************
ok: [vagrant-vm -> 127.0.0.1] => (item=~/.ssh/config)

TASK: [ansible-ssh | Update ssh config file] **********************************
failed: [vagrant-vm -> 127.0.0.1] => {"changed": false, "cmd": ["rsync", "-acv", "/tmp/ansible-ssh-config", "roles/ansible-ssh/files/config"], "delta": "0:00:00.031313", "end": "2015-08-27 20:57:59.052043", "rc": 12, "start": "2015-08-27 20:57:59.020730", "stdout_lines": ["building file list ... done"], "warnings": ["Consider using synchronize module rather than running rsync"]}
stderr: rsync: push_dir#3 "/Users/austinpray/Desktop/trellis/roles/ansible-ssh/files" failed: No such file or directory (2)
rsync error: errors selecting input/output files, dirs (code 3) at /SourceCache/rsync/rsync-45/rsync/main.c(580) [receiver=2.6.9]
rsync: connection unexpectedly closed (8 bytes received so far) [sender]
rsync error: error in rsync protocol data stream (code 12) at /SourceCache/rsync/rsync-45/rsync/io.c(453) [sender=2.6.9]
stdout: building file list ... done

PLAY [WordPress Server - Install LEMP Stack with PHP 5.6 and MariaDB MySQL] ***

GATHERING FACTS ***************************************************************
FATAL: no hosts matched or all hosts have already failed -- aborting


TASK: [common | Validate Ansible version] *************************************
FATAL: no hosts matched or all hosts have already failed -- aborting


PLAY RECAP ********************************************************************
           to retry, use: --limit @/Users/austinpray/dev.retry

vagrant-vm                 : ok=3    changed=0    unreachable=0    failed=1

Ansible failed to complete successfully. Any error output should be
visible above. Please fix these errors and try again.
```

austinpray commented:
```
diff --git a/roles/ansible-ssh/tasks/main.yml b/roles/ansible-ssh/tasks/main.yml
index bd5d04e..8768674 100644
--- a/roles/ansible-ssh/tasks/main.yml
+++ b/roles/ansible-ssh/tasks/main.yml
@@ -3,6 +3,10 @@
   local_action: file path=/tmp/ansible-ssh-config state=absent
   changed_when: false
 
+- name: Ensure ansible-ssh/files exists 
+  local_action: file path=roles/ansible-ssh/files/config state=directory
+  changed_when: false
+
 - name: Gather ssh config info for Vagrant VM
   local_action: shell vagrant ssh-config --host {{ inventory_hostname }} >> /tmp/ansible-ssh-config
   when: "'development' in group_names"
```

Works for me

austinpray commented:
>We don't have to rename the VM but it might be nice to have the names be consistent. Changing the name is a breaking change of sorts, but it seems better to do it now, while pre v1.0.

I'm good with this, `vagrant-vm` makes much more sense as to what the name is actually for. `default` is incredibly cryptic.

Also more errors:

![image](https://cloud.githubusercontent.com/assets/2192970/9537850/55005f5a-4d03-11e5-8487-7c5b7a311ed2.png)

This is after `vagrant destroy -f && vagrant up`. Investigating.

fullyint commented:
Thank you @austinpray. I've added your patch. On an earlier version I had committed a default config file to the repo, so its parent directory existed: `roles/ansible-ssh/files`. But I added the config file to the `.gitignore` so, then the `files` directory never made it into the repo. 

I've added the task you suggested, but immediately before the rsync task that would require the directory be present.

swalkinshaw commented on roles/ansible-ssh/tasks/main.yml:
```diff
@@ -0,0 +1,31 @@
+
+- name: Gather ssh config info for Vagrant VM
+  local_action: shell vagrant ssh-config --host {{ inventory_hostname }} >> /tmp/ansible-ssh-config
+  when: "'development' in group_names"
```
not needed?

fullyint replied:
This task must be ready to skip for staging/production hosts for whom we don't want to run `vagrant ssh-config`. I had to run this ansible-ssh role run in `server.yml` too because once `ansible.cfg` specifies a ssh config file, the file must be present and only that config file will be used. This doesn't mean you'll be without info from your regular `~/.ssh/config`, however, because this new config file will suck in whatever ssh config "includes" files you specify in `ssh_config_files` (defined in `group_vars/all/main.yml`).

fullyint commented:
**Problem:** The remote-user manual ssh test could inappropriately show that `root` cannot connect under this combination of conditions: 
* a host appears in the ansible hosts file as a nickname only, with no ansible_ssh_host (no IP)
* the host's IP is in a HostName parameter one of the `ssh_config_files` "includes", but not in the default `~/.ssh/config`
* the local action ssh test runs `ssh root@nickname` using default `~/.ssh/config` (no entry for nickname)
* connection fails with stderr `ssh: Could not resolve hostname nickname`

**Resolution:** I added a commit with two changes to resolve this.
* I moved the ansible-ssh role before the remote-user role. This assembles the `ssh_config_files` into one file at `roles/ansible-ssh/files/config`. Now the nickname host's `HostName` (IP) is available. 
* I adjusted the manual ssh `root` connection test to use this custom ssh config file by adding `-F roles/ansible-ssh/files/config`. Instead of adding `-F`, I'd rather just rebase this on top of #321 (if/when it's merged), which seems superior to the manual ssh check.

fullyint commented:
Closing. The proposed role is cool if and when you need it. However, only a small minority of users would need it. Anyone is welcome to turn it into a galaxy role.
